addons/account/models/account_payment_term.py

```diff
@@ -82,6 +99,38 @@ def unlink(self):
             ).unlink()
         return super(AccountPaymentTerm, self).unlink()
 
+    @api.onchange('discount_days')
+    def onchange_discount_days(self):
+        if self.discount_days <= 0:
+            raise ValidationError(_("The discount availability must be strictly positive."))
+        if self.discount_time_availability == "d_day_following_month" or self.discount_time_availability == "d_day_current_month":
+            if self.discount_days > 31:
+                raise ValidationError(_("This date is not valid."))
+
+    @api.onchange('discount_time_availability')
+    def onchange_discount_time_availability(self):
+        if self.discount_time_availability == "d_day_following_month" or self.discount_time_availability == "d_day_current_month":
+            if self.discount_days > 31:
+                raise ValidationError(_("This date is not valid."))
+
+    @api.onchange('percentage_to_discount')
+    def onchange_percentage_to_discount(self):
+        if self.percentage_to_discount <= 0:
+            raise ValidationError(_("The discount percentage must be strictly positive."))
+        if self.percentage_to_discount > 100:
+            raise ValidationError(_("The discount percentage cannot exceed 100."))
+
+    def get_last_date_for_discount(self, invoice_date):
+        if self.discount_time_availability == "d_day_after_invoice_date":
+            return invoice_date + timedelta(days=self.discount_days)
+        elif self.discount_time_availability == "d_after_invoice_month":
+            next_month = invoice_date.replace(day=28) + timedelta(days=4)
+            last_day = next_month - timedelta(days=next_month.day)
+            return last_day + timedelta(days=self.discount_days)
+        elif self.discount_time_availability == "d_day_following_month":
+            return (invoice_date.replace(day=1) + timedelta(days=32)).replace(day=self.discount_days)
+        elif self.discount_time_availability == "d_day_current_month":
+            return invoice_date.replace(day=self.discount_days)
```

Comment:
You probably need to handle the case when the discount_days is before/equal to the current date. In that case, you should go to the next month

Reply:
Agreed, though I think a ValidationError would maybe be a little clearer (_You cannot set a date equal to or before the invoice date_) for the user than taking the decision "behind his back" to move the day forward a month. Thoughts ?